Adding a hook (wrapper) for non-std stream reader in PyTorchStreamReader

houseroad opened the pull request:
To implement a stream is very annoying, since it is closely defined with the underlying storage streambuffer. 

So in this PR, we add ReadAdapterInterface and PyTorchStreamReader will use it. We implement IStreamAdapter as a wrapper of std::istream. And keep the user interface unchanged.

pritamdamania87 commented:
@houseroad I think you should be able to do it now. I added myself to the project.

zdevito commented on caffe2/serialize/istream_adapter.h:
```diff
@@ -0,0 +1,40 @@
+
+  size_t read(void* buf, size_t size) override {
+    istream_->read(static_cast<char*>(buf), size);
+    return istream_->gcount();
```
What is going on here? gcount was not previously called.

zdevito commented on caffe2/serialize/istream_adapter.h:
```diff
@@ -0,0 +1,40 @@
+ public:
+  explicit IStreamAdapter(std::istream* istream) : istream_(istream) {}
+
+  size_t size() override {
```
calling size changes the position of the stream!

zdevito commented on caffe2/serialize/read_adapter_interface.h:
```diff
@@ -0,0 +1,16 @@
+ public:
+  virtual size_t size() = 0;
+  virtual void seek(size_t pos) = 0;
+  virtual size_t read(void* buf, size_t size) = 0;
```
There is no need to have separate seek/read operators. Every time the container format reads it immediately seeks before hand. I think this should just be two functions:
`size_t read(uint64_t pos, char* buf, size_t n, const char* what)` and `size_t size()`.  `validate` is weird to have separate and can be folded into read.

zdevito commented on caffe2/serialize/inline_container.h:
```diff
@@ -109,15 +112,16 @@ class CAFFE2_API PyTorchStreamReader final {
+  std::unique_ptr<mz_zip_archive> ar_;
+  std::string archive_name_;
+  std::shared_ptr<at::ReadAdapterInterface> in_;
+  std::ifstream file_stream_;
```
Now that there is a third way to access the reader, it is messy to have ifstream initialized the way it is. It would make more sense to have a FileReader subclass of ReaderAdapterInterface of the reader interface be responsible for opening, reading, and closing the file if that constructor ran.

zdevito reviewed (commented):
I've added some suggestions to keep the interface clean. There is no reason to directly copy the istream interface as the adapter interface,  given that the zip archive already uses a much simpler interface.

dzhulgakov commented on caffe2/serialize/inline_container.cc:
```diff
@@ -42,27 +46,31 @@ static std::string basename(const std::string& name) {
+  return in_->read(pos, buf, n, "reading file");
+}
+
+PyTorchStreamReader::PyTorchStreamReader(std::string file_name, std::istream* in) {
```
btw, when `in` is passed, file_name is ignored. Should it be two separate constructors? (it was also in the original code)

zdevito replied:
Yeah, it can be deleted. An old version of the code needed to know the filename to know the top-level folder in archive where everything is, but now it just reads that from the first file in the archive.

dzhulgakov commented on caffe2/serialize/inline_container.cc:
```diff
@@ -42,27 +46,31 @@ static std::string basename(const std::string& name) {
-PyTorchStreamReader::PyTorchStreamReader(std::string file_name, std::istream* in)
-: ar_(new mz_zip_archive), in_(in) {
+PyTorchStreamReader::PyTorchStreamReader(std::shared_ptr<at::ReadAdapterInterface> in) {
+  in_ = in;
```
nit: put it into initialization list

smessmer replied:
and move it

houseroad replied:
changed to the unique_ptr already, and yes, moved it :-)

dzhulgakov commented on caffe2/serialize/inline_container.cc:
```diff
@@ -42,27 +46,31 @@ static std::string basename(const std::string& name) {
-    in_ = &file_stream_;
-    valid("opening archive");
-  }
+  valid("opening archive");
```
this one doesn't help any more (as `valid` only operates on miniz operations). remove?

dzhulgakov commented on caffe2/serialize/inline_container.cc:
```diff
@@ -191,10 +196,11 @@ size_t PyTorchStreamReader::getRecordOffset(const std::string& name) {
+      local_header,
+      MZ_ZIP_LOCAL_DIR_HEADER_SIZE,
+      "reading file header");
   valid("reading file header");
```
doesn't make sense any more

dzhulgakov commented on caffe2/serialize/istream_adapter.h:
```diff
@@ -0,0 +1,44 @@
+      return 0;
+    }
+    istream_->read(static_cast<char*>(buf), n);
+    if (!*istream_) {
```
same

dzhulgakov commented on caffe2/serialize/read_adapter_interface.h:
```diff
@@ -0,0 +1,14 @@
+// PyTorchStreamReader. with this interface, we can extend the support
+// besides standard istream
+class ReadAdapterInterface {
+ public:
```
add a virtual destructor and implementation in .cpp file please

dzhulgakov commented on caffe2/serialize/read_adapter_interface.h:
```diff
@@ -0,0 +1,14 @@
+class ReadAdapterInterface {
+ public:
+  virtual size_t size() = 0;
+  virtual size_t read(uint64_t pos, void* buf, size_t n, const char* what) = 0;
```
default `what` to nullptr?

houseroad replied:
I will just set the default value to "" (I feel it's safer)

houseroad commented on caffe2/serialize/istream_adapter.h:
```diff
@@ -0,0 +1,44 @@
+    return result;
+  }
+
+  size_t read(uint64_t pos, void* buf, size_t n, const char* what) override {
```
Yes, i am doing that.

houseroad commented on caffe2/serialize/istream_adapter.h:
```diff
@@ -0,0 +1,44 @@
+  explicit IStreamAdapter(std::istream* istream) : istream_(istream) {}
+
+  size_t size() override {
+    auto prev_pos = istream_->tellg();
```
Yeah, `validate` will catch it.

houseroad commented on caffe2/serialize/inline_container.cc:
```diff
@@ -42,27 +46,31 @@ static std::string basename(const std::string& name) {
+
+void PyTorchStreamReader::init() {
+  AT_ASSERT(in_ != nullptr);
+  ar_ = std::unique_ptr<mz_zip_archive>(new mz_zip_archive);
```
Sure, I will move it

smessmer replied:
prefer `make_unique`, never use `new` directly.

houseroad commented on caffe2/serialize/istream_adapter.h:
```diff
@@ -0,0 +1,44 @@
+  }
+
+ private:
+  std::istream* istream_;
```
I think let the caller to maintain the ownership is acceptable. Since this IStreamAdapter is only used for PyTorchStreamReader. I can change it to shared_ptr in a different PR, because it needs to change import_ir_module as well, which is out of the scope of this PR.

zdevito replied:
It is idiomatic to pass borrowed references to std::ostreams. Consider how operator<< typically works. It will be impossible to use with std::cout if this is changed to an std::shared_ptr, so please do not make this change.

pritamdamania87 commented on caffe2/serialize/istream_adapter.cpp:
```diff
@@ -0,0 +1,37 @@
+size_t
+IStreamAdapter::read(uint64_t pos, void* buf, size_t n, const char* what) {
+  istream_->seekg(pos);
+  validate(std::string("seeking, ").append(what).c_str());
```
How much do we care about efficiency of reading here? Each validate call is building a string and if we perform a lot of IStreamAdapter::read() calls, the validate would slow things down.

It seems like for large models this might be an issue. Do we need this function to take a `what` argument? Why don't we just have: `validate("seeking")` and `validate("reading")`

dzhulgakov replied:
it should be fine to just put `what` directly - I'm not even sure we need so detailed messages here

houseroad commented on caffe2/serialize/file_adapter.cpp:
```diff
@@ -0,0 +1,21 @@
+#include <caffe2/serialize/file_adapter.h>
+#include <caffe2/core/common.h>
+
+namespace at {
```
According to https://fb.quip.com/fNcfAiQ0pw1I, we are converging to `at` namespace, and this is a general class, which is not specific to jit module.

dzhulgakov replied:
I think we should put it into whatever namespace is the entire folder (right now torch::jit). cc @smessmer for opinion too

smessmer replied:
In c10, we decided that namespaces and module structure should (mostly) align. Applying that rule here would mean putting this into the `caffe2` namespace, or if the `caffe2/serialize` folder has enough files to warrant being its own module, it should be `caffe2::serialize`.

dzhulgakov reviewed (approved):
Looks good to me modulo the namespace discussion

smessmer commented on caffe2/serialize/read_adapter_interface.h:
```diff
@@ -0,0 +1,19 @@
+// besides standard istream
+class ReadAdapterInterface {
+ public:
+  virtual size_t size() = 0;
```
Is there a reason this isn't const? My first intuition would be to make both, `size` and `read` const since I wouldn't expect them to change state.

pritamdamania87 commented on caffe2/serialize/inline_container.cc:
```diff
@@ -42,27 +46,33 @@ static std::string basename(const std::string& name) {
-    valid("opening archive");
-  }
+PyTorchStreamReader::PyTorchStreamReader(std::istream* in)
+    : ar_(new mz_zip_archive),
```
You can use `make_unique` for mz_zip_archive too.

houseroad replied:
Yep, I should fix all the new.

smessmer commented on caffe2/serialize/inline_container.cc:
```diff
@@ -42,27 +46,33 @@ static std::string basename(const std::string& name) {
+
+PyTorchStreamReader::PyTorchStreamReader(
+    std::unique_ptr<at::ReadAdapterInterface> in)
+    : ar_(new mz_zip_archive), in_(std::move(in)) {
```
prefer `make_unique` to `new`. Also in other places in this diff.

pritamdamania87 reviewed (approved):
Looks good, although please address changes suggested by @smessmer

houseroad committed:
Adding a hook (wrapper) for non-std stream reader in PyTorchStreamReader (#15551)

Summary:
To implement a stream is very annoying, since it is closely defined with the underlying storage streambuffer.

So in this PR, we add ReadAdapterInterface and PyTorchStreamReader will use it. We implement IStreamAdapter as a wrapper of std::istream. And keep the user interface unchanged.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/15551

Differential Revision: D13568907

fbshipit-source-id: c807cff0799eb8d20ad7857ab592b9e10a7a7e61
